Upgrade elim-local-single-block for OpenCL.DebugInfo.100

greg-lunarg opened the pull request:
Essentially creates a DebugValue when removing a store to a local variable.

greg-lunarg commented:
This is WIP as I have not yet written a test. Once the reviewers have agreed to the approach I have used I will write a test.

The approach I have used is to generate DebugValue instructions without removing the associated DebugDeclare. I did not want to remove the DebugDeclare since this pass will always leave at least one Store to the variable.  So this pass will leave SPIR-V such that there are both DebugDeclares and DebugValues for the same variable.

This is not expressly ruled out in any spec I could find, although it probably isn't ideal for tools. The good news is that this is only done when it is quite certain that a subsequent pass of ssa-rewrite will clean up the DebugDeclare. Specifically, we will only replace a Store with a DebugValue for scalar variables.

For additional consideration, this pass was written solely to improve the performance of ssa-rewrite, so we can generally expect a pass of ssa-rewrite to follow. It does for all the optimization sequences currently used by glslang and dxc.

Also for consideration: ssa-rewrite and other analyses that remove stores should be able to deal with this hybrid situation without difficulty, just concentrating on the DebugDeclare and Stores and ignoring any associated DebugValues.

greg-lunarg commented:
@jaebaek Please review

jaebaek commented on source/opt/local_single_block_elim_pass.cpp:
```diff
@@ -73,11 +90,21 @@ bool LocalSingleBlockLoadStoreElimPass::LocalSingleBlockLoadStoreElim(
-                instructions_to_save.count(prev_store->second) == 0) {
+                instructions_to_save.count(prev_store->second) == 0 &&
+                !(IsArrayOrStruct(varId) &&
+                  get_debug_info_mgr()->IsDebugDeclared(varId))) {
```
In my opinion, we should allow delete store for this case. Consider the following example:
```
%empty_expr = OpExtInst %void %ext DebugExpression
...
%foo = OpVariable %ptr_Function_int Function
...
OpStore %foo %int_13  // will be deleted
OpStore %foo %int_17
%decl = OpExtInst %void %ext DebugDeclare %dbg_foo %foo %empty_expr
```
We have to delete `OpStore %foo %int_13`.

The following case looks non-trivial:
```
%empty_expr = OpExtInst %void %ext DebugExpression
...
%foo = OpVariable %ptr_Function_int Function
...
OpStore %foo %int_13  // will be deleted
%decl = OpExtInst %void %ext DebugDeclare %dbg_foo %foo %empty_expr
...
OpStore %foo %int_17
```

IMO, we have to change it to
```
%empty_expr = OpExtInst %void %ext DebugExpression
...
%foo = OpVariable %ptr_Function_int Function
...
%val = OpExtInst %void %ext DebugValue %dbg_foo %int_13 %empty_expr
...
OpStore %foo %int_17
%decl = OpExtInst %void %ext DebugDeclare %dbg_foo %foo %empty_expr
```

What do you think?

s-perron replied:
I think we should just not delete the store for any variable (scalars as well) that have a debug declare.  If DebugDeclare is suppose to match llvm.dbg.declare, the address has to always contains the correct value.  However, the spir-v extension is not as clear as the llvm-ir spec, this might be fine.  

I think that as long as we still remove the loads, this pass will still do what it needs to do.

greg-lunarg replied:
For clarification, do you believe that we should not remove the store even if we replace it with a DebugValue, which is being done here?

More generally, can we allow DebugDeclare and DebugValue to coexist for the same DebugLocalVariable? I think we can temporarily within spirv-opt as long as we ultimately remove the DebugDeclares before outputting the SPIR-V.

s-perron replied:
I don't think we should remove the store.  The DebugValue is not needed, but I don't think it hurts.

> can we allow DebugDeclare and DebugValue to coexist for the same DebugLocalVariable?

The spec is so unclear, I am not sure what we can or cannot do.  I believe it would be okay to have both as long as they give the same information.

>  I think we can temporarily within spirv-opt as long as we ultimately remove the DebugDeclares before outputting the SPIR-V.

If we could guarantee that users always run ssa-rewrite then I could agree with this: temporarily invalid code is fine.  However, our passes are meant to be used in potentially any order.  We have tried to ensure that no pass requires another pass to run in order to be correct.  The only exception so far is merge-return and dead-branch elimination.  In that case an error is emitted (https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/opt/merge_return_pass.cpp#L108).

jaebaek replied:
Ah ok, my comment was wrong.
[llvm-dbg-declare](https://llvm.org/docs/SourceLevelDebugging.html#llvm-dbg-declare) explains "It is not control-dependent ... that address is considered to be the true home of the variable across its entire lifetime."

We must not delete the store if the variable is used for `DebugDeclare`.

greg-lunarg replied:
Not sure why you think we should not remove the OpStore.

Do we agree that for a scalar local (function scope) OpVariable with a DebugLocalVariable and a DebugDeclare, it is ok , and actually a good thing (a goal for spirv-opt) to replace all its OpStores with DebugValues, remove the DebugDeclare and the actual OpVariable, but leave the DebugLocalVariable in place?

What I would like to establish is what we want the legalization passes to do and not do in the presense of OpenCL.DebugInfo.100.

Can/should the legalization passes do less optimization in the presense of OpenCL.DebugInfo.100? For example, can/should the passes not eliminate OpStores? I think it is possible to eliminate OpStores as long as we replace them with DebugValues.

greg-lunarg replied:
I think it might be possible to legalize without removing OpStores. I will have to go back to our original legalization shaders and see. I just thought, to be safe, we should try to stick to the original passes as closely as possible but still retain the debug information.

greg-lunarg replied:
I am asking because I am looking at DCE right now which eliminates a lot of OpStores. If you think that is bad, it would be good to know now.

s-perron replied:
> Not sure why you think we should not remove the OpStore.

If you look up the spec for llvm.dbg.declar (which Jaebaek referenced above), it clearly says that if you have an llvm.dbg.declare for a local variable, then that is the address for the variable for its entire lifetime.  If we remove a store, while leaving a debug declare, then that will have been violated.  That is why I don't think we can remove the store.

s-perron replied:
> Do we agree that for a scalar local (function scope) OpVariable with a DebugLocalVariable and a DebugDeclare, it is ok , and actually a good thing (a goal for spirv-opt) to replace all its OpStores with DebugValues, remove the DebugDeclare and the actual OpVariable, but leave the DebugLocalVariable in place?

Yes, I agree.  The ssa-rewriter should be able to do that, and it will get done.  The ssa-rewritter will replace the DebugDeclare with all of the apporpriate DebugValue instruction, and then we can remove the stores.  If the DebugDeclare is still present, we cannot remove the stores.

s-perron replied:
> Can/should the legalization passes do less optimization in the presense of OpenCL.DebugInfo.100? For example, can/should the passes not eliminate OpStores? I think it is possible to eliminate OpStores as long as we replace them with DebugValues.

If the OpVariable is used by a DebugDeclare or a DebugValue then we should not remove the store unless you can also remove the use in the DebugDeclare or DebugValue.  ssa-rewrite if removing the DebugDeclare and DebugValue, so later passes like ADCE can remove the stores.  We can still remove the loads at any time, and do the copy propagation etc.

s-perron replied:
> I think it might be possible to legalize without removing OpStores. I will have to go back to our original legalization shaders and see. I just thought, to be safe, we should try to stick to the original passes as closely as possible but still retain the debug information.

I believe so as well, as long at the loads are removed.  That is what we really wanted from elim-local-single-block and elim-local-single-store anyway.

s-perron replied:
> I am asking because I am looking at DCE right now which eliminates a lot of OpStores. If you think that is bad, it would be good to know now.

As I said before DCE should not remove the store if the OpVariable is used in a DebugDeclare or DebugValue.  The ssa rewriter should remove those uses in cases where we can remove the stores, so I don't think it will hinder optimization.

greg-lunarg replied:
I think it is ok to replace a single OpStore with a DebugValue as long as we eventually replace them all and get rid of the DebugDeclare before outputting the SPIR-V.

I agree with the general philosophy that each pass should be able to stand alone, but elim-local-single-block is not really a stand-alone pass; it was written to be a helper to ssa-rewrite, so I think it is ok to assume that ssa-rewrite will follow.

That said, I don't think its a huge deal if we don't eliminate OpStores in elim-local-single-block. I was just trying to create as little new logic and exceptions and different behavior for DebugInfo as possible in the interest of reducing complexity.

greg-lunarg replied:
OK. Just looked at output from ssa-rewrite. Now I understand your approach better.

ssa-rewrite is adding DebugValue after each Store and removing the DebugDeclare. So essentially ssa-rewrite is preparing the OpStores and OpVariable for deletion.

I can work with this model now that I understand it.

s-perron commented:
Thanks for this Greg.  This looks good in general.  The problem I am having with all of this debug stuff is how imprecise the specification of the debug extension is.

greg-lunarg commented:
I have changed this to only avoiding deleting a store in the case that the variable is has a DebugDeclare, deferring to ssa-rewrite to take care of it. I have also added a test and thus have removed the WIP.

s-perron commented on source/opt/local_single_block_elim_pass.cpp:
```diff
@@ -31,6 +31,10 @@ const uint32_t kStoreValIdInIdx = 1;
   if (supported_ref_ptrs_.find(ptrId) != supported_ref_ptrs_.end()) return true;
   if (get_def_use_mgr()->WhileEachUser(ptrId, [this](Instruction* user) {
+        auto dbg_op = user->GetOpenCL100DebugOpcode();
+        if (dbg_op == OpenCLDebugInfo100DebugDeclare) {
```
@greg-lunarg I believe that jaebaek is right.  You should look for a DebugValue here as well.  Scalar replacement will be replacing DebugDeclare instructions with DebugValue instruction as shown above.  This would still be a supported ref.

s-perron commented on test/opt/local_single_block_elim.cpp:
```diff
@@ -1118,6 +1118,182 @@ TEST_F(LocalSingleBlockLoadStoreElimTest, VariablePointerTest) {
   SinglePassRunAndMatch<LocalSingleBlockLoadStoreElimPass>(text, false);
 }
+
+TEST_F(LocalSingleBlockLoadStoreElimTest, DebugInfoTest) {
```
This is a good test.  We should add another one where a DebugValue with a dref expression is used in place of the DebugDeclare.  We would expect the same resutls.

greg-lunarg replied:
I admit that the OpenCL.DebugInfo.100 spec is vague, but after some research and thought, I do not believe that a DebugValue with a deref of a pointer Value is the same as a DebugDeclare. A DebugValue represents an assignment of the Value and Expression to the **rvalue** of the LocalVariable. A DebugDeclare represents an "equivalencing" of the Expression applied to lvalue of the Variable to the **lvalue** of the LocalVariable. I am fairly certain these are not the same thing.

I am also fairly certain that we will never have cause to have a deref in an Expression in a DebugValue in a shader. The Variable would have to be of type pointer into function scope memory which don't exist in shaders.

greg-lunarg replied:
After some thought, while I am still fairly certain that one cannot use a DebugValue to create an equivalent to a DebugDeclare in a shader, I will grant that one could create a legitimate DebugValue that contains a deref of a local pointer for a shader, However, I think that such a reference would be so highly unlikely that it would not be worth creating the logic to add it to the supported references for this optimization. If such a thing did appear, it would merely disable optimization and not generate incorrect code. So I think the existing logic is reasonable and acceptable at this time. If someone can make an argument for why such a construct would be prevalent enough to justify optimization here, we can reconsider this logic at that time.

s-perron replied:
> A DebugValue represents an assignment of the Value and Expression to the rvalue of the LocalVariable.

I have thought about this a lot as well when trying to figure out the design for scalar replacement.  I understand your point of view.  At first, I was thinking that way as well.  However, I started thinking about how this would actually be implemented by a debugger, and in dwarf.  In DWARF, you don't give the value of a variable.  DWARF say where you go to get the value (register or memory).  The important point to realize is that the DebugValue is not an assignment of a value.  This can be easily confused when rerefs are not used because, in llvm-ir and in spir-v, the ssa value cannot be changed.

The other reason that I believed the llvm.dbg.value could not be viewed as an assignment of a value is the extra work this would put on debuggers, and how it would slow down debugging.  The debugging info is not present in the machine code.  If the debugger was required to capture the value of the expression at every llvm.dbg.value, the debugger would have to stop the normal execution of the program at every position of an llvm.dbg.value to read the value.  This would cause a ridiculous amount of overhead in debugging.  We wouldn't want a spir-v debugger to have to do the same.

Granted this was just guesses, which is why I ask Jaebaek to look at what code actually gets generated by llvm-ir when a debug value is used with a dref expression, and to examine what a real debugger does with it.  We hand wrote llvm-ir that has exactly one llvm.dbg.value where the address of the memory location is the value and there is a deref expression.  The instruction was placed after the first store to the memory location.  There was a second store to the same memory location, but there was not a second llvm.dbg.value.  When running in the debugger, you see the value of the first store when you break after the first store.  However, when you break after the second store, you see the result of the second store.  The debugger did not capture the value of the first store to be used anytime the developer asks for the value of the variable.  The debugger checks the value in the memory location.

greg-lunarg replied:
OK. Thanks for the detailed response. I do now agree that a DebugDeclare can be replaced with a DebugValue with a deref, and I apologize that it took me this long to come to this understanding.

So I guess now the question in my mind is: is this really what we want to do?

Essentially, to do this, we have to keep backing memory for all locals (function scope variables) and keep all Stores to that memory that we normally remove. So we will have to add extra logic to all optimizations that eliminate stores. In one way it will be simpler for the optimizer since it only has to generate one DebugValue at the top of the program, but there will be performance cost from the extra Stores.

This design will also require a whole additional and redundant memory analysis of derefs in DebugValues.

The next question is: do we need to do this? You opined above that DebugValue would be intractable for debuggers as I imagined them. But if you look at https://llvm.org/docs/SourceLevelDebugging.html#object-lifetime-in-optimized-code it appears that llvm.dbg.value is being used in just this manner.

As far as debugger implementation, I imagine the debugger would be able to create a fairly simple map for each local variable, mapping a line number range to a register.

Still, your design will functionally work, albeit with probable performance costs, and I can work with it.

greg-lunarg replied:
I think I can just look at the Value field of the DebugValue and see if its a pointer, so that should be fairly easy.

jaebaek replied:
Your questions are very important questions to remove the technical debt and make a reasonable design. I appreciate it. I am conducting more experiments to figure out the relationship between `llvm.dbg.declare` and `llvm.dbg.value` particularly with `DW_OP_LLVM_fragment`. I am also checking `DW_OP_bit_piece`, which is similar to `DW_OP_LLVM_fragment`.

IIRC, we decided to use `DebugValue` with `Deref` to express the declaration because we thought it is the only way to express the declaration in the scalar replacement. In LLVM world, SROA conducts both the scalar replacement and ssa-rewrite in a single pass, so it does not have to consider "splitting a composite type while keeping the information of declaration". On the other hand, spirv-opt scalar replacement must "split a composite type while keeping the information of declaration". Since `DebugDeclare` does not have `Indexes` operand while `DebugValue` does, we had no idea how to use `DebugDeclare` for each member of a composite. See [DebugDeclare](https://www.khronos.org/registry/spir-v/specs/unified1/OpenCL.DebugInfo.100.html#DebugDeclare).

The main questions I currently have are
- **Q1**: Does LLVM SROA pass propagate `llvm.dbg.value` for store and phi instructions when we give `llvm.dbg.value` with `DW_OP_deref`?
- **Q2**: Can we use `DW_OP_LLVM_fragment` with `llvm.dbg.declare` for the scalar replacement?
- **Q3**: Is there a way to express a member of a composite using `DebugDeclare` and `Deref` operation?

If **Q2** and **Q3** are true, I think we do not have to use `DebugValue` with `Deref` for the declaration in the scalar replacement.
In addition, if **Q2** and **Q3** are true and **Q1** is false, it would be better to follow what LLVM does i.e., we must not use `DebugValue` with `Deref` as a declaration.

In my brief experiments, **Q1** and **Q2** are true, but I am doublechecking it if I made some mistakes. I am also digging into LLVM source code.
As far as I understand, `DW_OP_LLVM_fragment` is applied to the variable not its value of `llvm.dbg.value` nor its storage of `llvm.dbg.declare`.
I am curious what `DW_OP_bit_piece` means because I guess (I do not have any evidence yet) that it is applied to the value of `llvm.dbg.value` or the storage of `llvm.dbg.declare`.

For **Q3**, I am discussing the spec change with the initial author of the spec.

greg-lunarg commented:
@s-perron I think have provided all of your requested changes. Is there anything else that is needed?

jaebaek commented:
@greg-lunarg Steven is on vacation and he will not work for a while. I think this PR is ready to be submitted.

I will open a separate bug if we have to update the part related to `DebugValue`.

s-perron commented:
This looks good to me.  Thanks.
